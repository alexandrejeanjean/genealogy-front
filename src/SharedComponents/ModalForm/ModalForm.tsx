import React, { PureComponent } from "react";
import { Form, Button, Modal } from "react-bootstrap";

type TInputs = {
  name: string;
  placeholder: string;
  inputType: string;
  datas?: Array<any>;
};

type Props = {
  title: string;
  inputs: TInputs[];
  submit: Function;

  show: boolean;
  onHide: Function;
};

interface State {
  name: string;
  firstname: string;
  lastname: string;
  position: number | null;
  role: string;
}

class ModalForm extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      name: "",
      position: null,
      firstname: "",
      lastname: "",
      role: "",
    };
  }

  // Simple inputs verificator: will improve this with a json-validator
  checkDatas = () => {
    const { name, role, firstname, lastname, position } = this.state;
    const { title } = this.props;

    if (["Person"].indexOf(title) > -1) {
      if (role === "" || firstname === "" || lastname === "") {
        return true;
      }
    } else if (["Family"].indexOf(title) > -1) {
      if (name === "") {
        return true;
      }
    } else if (["Generation"].indexOf(title) > -1) {
      if (!position) {
        return true;
      }
    } else {
      return false;
    }
  };

  handleChange = (event: React.FormEvent<{ name: string; value: string }>) => {
    const name = event.currentTarget.name;
    const value = event.currentTarget.value;
    this.setState(({ [name]: value } as unknown) as Pick<State, keyof State>);
  };

  handleSubmit = (event: React.FormEvent<HTMLInputElement>) => {
    event.preventDefault();
    const { onHide, submit } = this.props;
    submit(this.state);
    onHide();
  };

  render() {
    const { title, inputs, onHide, show } = this.props;

    return (
      <Modal
        size="lg"
        aria-labelledby="contained-modal-title-vcenter"
        centered
        onHide={onHide}
        show={show}
      >
        <Modal.Header closeButton>
          <Modal.Title id="contained-modal-title-vcenter">
            New {title}
          </Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Form onSubmit={this.handleSubmit}>
            {inputs.map((input: TInputs) => (
              <Form.Group
                controlId={"formControl_" + input.name}
                key={input.placeholder}
              >
                <Form.Label>
                  {input.name.charAt(0).toUpperCase() + input.name.substr(1)}
                </Form.Label>
                {input.inputType === "select" ? (
                  <Form.Control
                    as={input.inputType}
                    name={input.name}
                    onChange={this.handleChange}
                    defaultValue={"DEFAULT"}
                  >
                    <option value="DEFAULT" disabled>
                      Select the role
                    </option>
                    {input.datas?.map((data) => (
                      <option key={data.role}>{data.role}</option>
                    ))}
                  </Form.Control>
                ) : (
                  <Form.Control
                    type={input.inputType}
                    name={input.name}
                    placeholder={input.placeholder}
                    onChange={this.handleChange}
                  />
                )}
              </Form.Group>
            ))}

            <Button
              type="submit"
              className="btn-primary mt-3"
              disabled={this.checkDatas()}
            >
              Create {title}
            </Button>
          </Form>
        </Modal.Body>
      </Modal>
    );
  }
}

export default ModalForm;
