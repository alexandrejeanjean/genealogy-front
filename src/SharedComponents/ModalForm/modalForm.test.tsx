import React from "react";
import { shallow, mount } from "enzyme";
import ModalForm from "./ModalForm";

const modalShow = true;
const setModalShow = jest.fn();
const createFamily = jest.fn();

const inputsArr = [
  { name: "name", placeholder: "Ex: Armstrong", inputType: "text" },
];
describe("ModalForm", () => {
  it("should render correctly", () => {
    const Component = shallow(
      <ModalForm
        title="Family"
        show={modalShow}
        inputs={inputsArr}
        onHide={() => setModalShow(false)}
        submit={createFamily}
      />
    );
    expect(Component).toMatchSnapshot();
  });

  it("should render form with the right number of inputs passed in props", () => {
    const Component = mount(
      <ModalForm
        title="Family"
        show={modalShow}
        inputs={inputsArr}
        onHide={() => setModalShow(false)}
        submit={createFamily}
      />
    );
    expect(Component.find(".form-group input").length).toBe(inputsArr.length);
    Component.unmount();
  });
});
