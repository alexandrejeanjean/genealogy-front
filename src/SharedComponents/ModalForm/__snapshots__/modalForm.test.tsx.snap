// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalForm should render correctly 1`] = `
<Modal
  animation={true}
  aria-labelledby="contained-modal-title-vcenter"
  autoFocus={true}
  backdrop={true}
  centered={true}
  dialogAs={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "displayName": "ModalDialog",
      "render": [Function],
    }
  }
  enforceFocus={true}
  keyboard={true}
  onHide={[Function]}
  restoreFocus={true}
  show={true}
  size="lg"
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      id="contained-modal-title-vcenter"
    >
      New 
      Family
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    <Form
      inline={false}
      onSubmit={[Function]}
    >
      <FormGroup
        controlId="formControl_name"
        key="Ex: Armstrong"
      >
        <FormLabel
          column={false}
          srOnly={false}
        >
          Name
        </FormLabel>
        <FormControl
          name="name"
          onChange={[Function]}
          placeholder="Ex: Armstrong"
          type="text"
        />
      </FormGroup>
      <Button
        active={false}
        className="btn-primary mt-3"
        disabled={true}
        type="submit"
        variant="primary"
      >
        Create 
        Family
      </Button>
    </Form>
  </ModalBody>
</Modal>
`;
